src/java.base/share/classes/java/util/Currency.java

```diff
@@ -432,61 +433,82 @@ public static Currency getInstance(Locale locale) {
-                    for (char c2 = 'A'; c2 <= 'Z'; c2 ++) {
-                        int tableEntry = getMainTableEntry(c1, c2);
-                        if ((tableEntry & COUNTRY_TYPE_MASK) == SIMPLE_CASE_COUNTRY_MASK
-                             && tableEntry != INVALID_COUNTRY_ENTRY) {
-                            char finalChar = (char) ((tableEntry & SIMPLE_CASE_COUNTRY_FINAL_CHAR_MASK) + 'A');
-                            int defaultFractionDigits = (tableEntry & SIMPLE_CASE_COUNTRY_DEFAULT_DIGITS_MASK) >> SIMPLE_CASE_COUNTRY_DEFAULT_DIGITS_SHIFT;
-                            int numericCode = (tableEntry & NUMERIC_CODE_MASK) >> NUMERIC_CODE_SHIFT;
-                            StringBuilder sb = new StringBuilder();
-                            sb.append(c1);
-                            sb.append(c2);
-                            sb.append(finalChar);
-                            available.add(getInstance(sb.toString(), defaultFractionDigits, numericCode));
-                        } else if ((tableEntry & COUNTRY_TYPE_MASK) == SPECIAL_CASE_COUNTRY_MASK
-                                && tableEntry != INVALID_COUNTRY_ENTRY
-                                && tableEntry != COUNTRY_WITHOUT_CURRENCY_ENTRY) {
-                            int index = SpecialCaseEntry.toIndex(tableEntry);
-                            SpecialCaseEntry scEntry = specialCasesList.get(index);
-
-                            if (scEntry.cutOverTime == Long.MAX_VALUE
-                                    || System.currentTimeMillis() < scEntry.cutOverTime) {
-                                available.add(getInstance(scEntry.oldCurrency,
-                                        scEntry.oldCurrencyFraction,
-                                        scEntry.oldCurrencyNumericCode));
-                            } else {
-                                available.add(getInstance(scEntry.newCurrency,
-                                        scEntry.newCurrencyFraction,
-                                        scEntry.newCurrencyNumericCode));
-                            }
+        initAvailableCurrencies();
+        return new HashSet<>(available);
+    }
+
+    /**
+     * {@return a stream of available currencies} The returned stream of currencies
+     * contains all the available currencies, which may include currencies
+     * that represent obsolete ISO 4217 codes. If there is no currency
+     * available in the runtime, the returned stream is empty.
+     *
+     * @implNote Unlike {@link #getAvailableCurrencies()}, this method does
+     * not create a defensive copy of the {@code Currency} set.
+     * @see #getAvailableCurrencies()
+     * @since 25
+     */
+    public static Stream<Currency> availableCurrencies() {
+        initAvailableCurrencies();
+        return available.stream();
+    }
+
+    // Initialize the set of available currencies if needed
+    private static synchronized void initAvailableCurrencies() {
```

Comment:
Should we make `available` volatile so we can avoid entering the monitor if the currencies are initialized?

(Note: We may get Stable Values very soon, at that point SV is a much better solution)

Reply:
Thanks Chen, made `available` volatile and implemented DCL in https://github.com/openjdk/jdk/pull/23165/commits/96e86c800f1d1af1c20bfee944c298f17b7e8860.
Also added a duplicate elements test as you suggested.

Reply:
I'm not sure how useful it is to optimize the performance of `availableCurrencies` access to "available", but adding volatile will slow every access down.
The computation of the available currencies is stable, so a race computing it is benign.
Except for the available hashSet being partially filled when read by the thread calling `getAvailableCurrencies()`.
That could be remedied by using a new local for the new HashSet in initAvailableCurrencies and assigning to available only when the set is completely initialized.
(And yes, this looks like a good fit for stable values.)

Reply:
How about I revert the most recent commit and have this PR simply deliver the change to provide the stream of currencies. I'll file another issue to improve `available` to use stable values when that is available. Does that sound fine?

Reply:
Reverted in https://github.com/openjdk/jdk/pull/23165/commits/ade21c923369d0f0423335c3073e2c615adf050c. Filed [JDK-8348351](https://bugs.openjdk.org/browse/JDK-8348351) to implement `available` as a stable value when applicable.

Reply:
I suspect that [Stable Values](https://openjdk.org/jeps/502) would fit well here.